client/asset/eth/txdb.go

```diff
@@ -0,0 +1,551 @@
+
+			opts := badger.DefaultIteratorOptions
+			it := txn.NewIterator(opts)
+			defer it.Close()
+
+			for it.Rewind(); it.Valid(); it.Next() {
+				item := it.Item()
+				key := item.Key()
+				newKey := monitoredTxKey(key)
+				monitoredTxB, err := item.ValueCopy(nil)
+				if err != nil {
+					return err
+				}
+
+				err = txn.Set(newKey, monitoredTxB)
+				if err != nil {
+					return err
+				}
+				err = txn.Delete(key)
+				if err != nil {
+					return err
+				}
+			}
+
+			versionB := make([]byte, 8)
+			binary.BigEndian.PutUint64(versionB, 1)
+			return txn.Set(dbVersionKey, versionB)
+		})
+	} else if version > txDBVersion {
+		return fmt.Errorf("database version %d is not supported", version)
+	}
+
+	return nil
+}
+
+func (s *badgerTxDB) close() error {
+	return s.Close()
+}
+
+// storeTx stores a mapping from nonce to extendedWalletTx and a mapping from
+// transaction hash to nonce so transactions can be looked up by hash. If a
+// nonce already exists, the extendedWalletTx is overwritten.
+func (s *badgerTxDB) storeTx(nonce uint64, wt *extendedWalletTx) error {
+	wtB, err := json.Marshal(wt)
+	if err != nil {
+		return err
+	}
+	nk := nonceKey(nonce)
+	tk := txHashKey(wt.ID)
+
+	return s.Update(func(txn *badger.Txn) error {
+		oldWtItem, err := txn.Get(nk)
+		if err != nil && !errors.Is(err, badger.ErrKeyNotFound) {
+			return err
+		}
+
+		// If there is an existing transaction with this nonce, delete the
+		// mapping from tx hash to nonce.
+		if err == nil {
+			oldWt := new(extendedWalletTx)
+			err = oldWtItem.Value(func(oldWtB []byte) error {
+				err := json.Unmarshal(oldWtB, oldWt)
+				if err != nil {
+					s.log.Errorf("unable to unmarhsal wallet transaction: %s: %v", string(oldWtB), err)
+				}
+				return err
+			})
+			if err == nil && !bytes.Equal(oldWt.ID, wt.ID) {
+				err = txn.Delete(txHashKey(oldWt.ID))
+				if err != nil {
+					s.log.Errorf("failed to delete old tx id: %s: %v", oldWt.ID.String(), err)
+				}
+			}
+		}
+
+		// Store nonce key -> wallet transaction
+		if err := txn.Set(nk, wtB); err != nil {
+			return err
+		}
+
+		// Store tx hash -> nonce key
+		return txn.Set(tk, nk)
+	})
+}
+
+// removeTx removes a tx from the db.
+func (s *badgerTxDB) removeTx(id dex.Bytes) error {
+	tk := txHashKey(id)
+
+	return s.Update(func(txn *badger.Txn) error {
+		txIDEntry, err := txn.Get(tk)
+		if err != nil {
+			return err
+		}
+
+		err = txn.Delete(tk)
+		if err != nil {
+			return err
+		}
+
+		nk, err := txIDEntry.ValueCopy(nil)
+		if err != nil {
+			return err
+		}
+
+		nkEntry, err := txn.Get(nk)
+		if err != nil {
+			return nil
+		}
+
+		wtB, err := nkEntry.ValueCopy(nil)
+		if err != nil {
+			return err
+		}
+
+		wt := new(extendedWalletTx)
+		err = json.Unmarshal(wtB, wt)
+		if err != nil {
+			s.log.Errorf("unable to unmarhsal wallet transaction: %s: %v", string(wtB), err)
+			return err
+		}
+
+		if bytes.Equal(wt.ID, id) {
+			return txn.Delete(nk)
+		}
```

Comment:
I was checking to see if the `walletTransaction` that the nonce mapped to contained the same ID was the one we were deleting. This is unnecessary though, because expired `txID -> nonce` mappings are being deleted in `storeTx`.